File: 2015/cpp/day01/main.cpp
#include <cassert>
#include <string>
#include <utility>
#include <iostream>
#include <fstream>

namespace {

int WhatFloor(const std::string &s) {
    int floor = 0;
    for (char c : s) {
        switch (c) {
        case '(':
            ++floor;
            break;
        case ')':
            --floor;
            break;
        default:
            std::unreachable();
            break;
        }
    }

    return floor;
}

int FindFirstBasement(const std::string &s) {
    int floor = 0;
    int i = 1;
    for (char c : s) {
        switch (c) {
        case '(':
            ++floor;
            break;
        case ')':
            --floor;
            break;
        default:
            std::unreachable();
            break;
        }

        if (floor == -1) {
            return i;
        }

        ++i;
    }

    return 0;
}

void Test() {
    // Problem 1
    assert(WhatFloor("(())") == 0);
    assert(WhatFloor("()()") == 0);
    assert(WhatFloor("(((") == 3);
    assert(WhatFloor("(()(()(") == 3);
    assert(WhatFloor("))(((((") == 3);
    assert(WhatFloor("())") == -1);
    assert(WhatFloor("))(") == -1);
    assert(WhatFloor(")))") == -3);
    assert(WhatFloor(")())())") == -3);

    // Problem 2
    assert(FindFirstBasement(")") == 1);
    assert(FindFirstBasement("()())") == 5);
}

} // namespace

int main() {
    Test();

    std::ifstream file("../../input/day01.txt");
    std::string input;
    file >> input;

    int ret1 = WhatFloor(input);
    int ret2 = FindFirstBasement(input);

    std::cout << "Problem1 = " << ret1 << std::endl;
    std::cout << "Problem2 = " << ret2 << std::endl;

    return 0;
}

File: 2015/rust/day01/src/main.rs
use std::fs;

fn what_floor(s: &str) -> i32 {
    s.chars()
        .fold(0, |acc, c| if c == '(' { acc + 1 } else { acc - 1 })
}

fn find_first_basement(s: &str) -> usize {
    let mut floor = 0;
    for (i, c) in s.chars().enumerate() {
        floor += if c == '(' { 1 } else { -1 };
        if floor == -1 {
            return i + 1;
        }
    }

    unreachable!();
}

fn main() {
    let input = fs::read_to_string("../../input/day01.txt").unwrap();
    let ret1 = what_floor(&input);
    let ret2 = find_first_basement(&input);

    println!("ret1={ret1}(expected: 232)");
    println!("ret2={ret2}(expected: 1783)");
}

#[test]
fn test1() {
    assert_eq!(what_floor("(())"), 0);
    assert_eq!(what_floor("()()"), 0);
    assert_eq!(what_floor("((("), 3);
    assert_eq!(what_floor("(()(()("), 3);
    assert_eq!(what_floor("())"), -1);
    assert_eq!(what_floor(")))"), -3);
    assert_eq!(what_floor(")())())"), -3);
}

#[test]
fn test2() {
    assert_eq!(find_first_basement(")"), 1);
    assert_eq!(find_first_basement("()())"), 5);
}
